### ISSUE TITLE
OpenShift routes for KServe InferenceServices

### ISSUE THREAD

**Issue by israel-hdez** [opendatahub-io/odh-model-controller#59]
In KServe, the URL where an InferenceService can be reached is reported in the same InferenceService. For example, this is a sample output when querying InferenceServices:

```
$ oc get isvc -A
NAMESPACE            NAME                      URL                                                                            READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                       AGE
kserve-ci-e2e-test   isvc-sklearn-v2-mixed     http://isvc-sklearn-v2-mixed-kserve-ci-e2e-test.apps.openshift-cluster.com     True           100                              isvc-sklearn-v2-mixed-predictor-00001     6d6h
kserve-ci-e2e-test   isvc-tensorflow-runtime   http://isvc-tensorflow-runtime-kserve-ci-e2e-test.apps.openshift-cluster.com   True           100                              isvc-tensorflow-runtime-predictor-00001   6d6h
```

The issue is that the specified URL is not reachable in OpenShift.

**Describe the solution you'd like**

KServe assumes a wildcard DNS configuration, as mentioned here: 
* https://knative.dev/docs/install/yaml-install/serving/install-serving-with-yaml/#configure-dns (See _Real DNS_ tab)

Typically, in OpenShift, a similar DNS configuration is already set. However, while KServe assumes that the DNS configuration points to Istio Ingress Gateway, in OpenShift the DNS configuration points to OpenShift's Ingress.  This is the reason of seeing the generic "Application is not available" page when trying to access the URLs specified on the InferenceServices.

Since the URLs are hitting OpenShift's Ingress, it turns out that it is a matter of creating the right OpenShift Route, in order to properly enable the InferenceService endpoints. So, **the goal in this ticket is to create these missing Route resources**.

For example, the following route will make reachable the InferenceService `isvc-sklearn-v2-mixed` that is present in the example above:

```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  name: isvc-sklearn-v2-mixed-kserve-ci-e2e-test
  namespace: istio-system
spec:
  host: isvc-sklearn-v2-mixed-kserve-ci-e2e-test.apps.openshift-cluster.com
  port:
    targetPort: https
  tls:
    insecureEdgeTerminationPolicy: Redirect
    termination: passthrough
  to:
    kind: Service
    name: istio-ingressgateway
    weight: 100
  wildcardPolicy: None
```

NOTE that:
* The target service is the Istio ingress gateway. This is fixed.
* The Routes should always be created in the Service Mesh control plane namespace, because that's the namespace where the Istio Ingress Gateway lives.
* In the provided examples, the Route is an HTTPS route, but the InferenceService displays a plain HTTP URL. The protocol at the InferenceService should be honored when creating the Route.
* The Route should NOT be created if the URL of the InfrenceService is an "in-cluster URL".
* The Route should be created only for KServe InferenceServices - not ModelMesh, where there is already a different logic.

**Comment by vaibhavjainwiz** [comment]
As I analysed there are two ways to fix this issue:
 - **Disable virtual host functionality of kserve** : If we set  `"disableIstioVirtualHost": true` inside 
inferenceservice-config configmap then predictor route would be set as exposed URL for InferenceService.
**Pros** : No code change required. Just some configuration changes in ConfigMap.
**Cons** : It would directly expose predictor route to user. Exposed URL would only work for predictor component. Exposed URL would not work for transformer.

- **Create generic openshift route** : Create generic openshift route which work for both predictor and transformer.
**Pros** : It work for both Predictor and Transformer.
**Cons** : Need code changes.

**Comment by vaibhavjainwiz** [comment]
As discussed in team meeting today, for 1.32 we would disable the virtual host functionality of kserve so that predictor route url would be exposed to user. For long term, it is recommended to create generic openshift route.

**Comment by vaibhavjainwiz** [comment]
**Testing Instruction :**
1. Create an InferenceService
2. run `oc get InferenceService`
3. run prediction using exposed url.

**Comment by vaibhavjainwiz** [comment]
PR : https://github.com/opendatahub-io/kserve/pull/63, https://github.com/opendatahub-io/kserve/pull/64, https://github.com/opendatahub-io/odh-manifests/pull/916

**Comment by heyselbi** [comment]
Upstream is merged https://github.com/opendatahub-io/odh-manifests/pull/916
Downstream bits are merged https://github.com/red-hat-data-services/odh-manifests/pull/472, https://github.com/red-hat-data-services/odh-manifests/pull/471

**Comment by israel-hdez** [comment]
We probably should have a follow-up on this, as I understood that for next version we should aim at creating the routes.

**Comment by bdattoma** [comment]
is this related to the steps reported in these instructions? https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/demo/kserve/Kserve.md
I'm referring to these lines:
```
# Generate wildcard cert for a gateway.
export BASE_DIR=/tmp/kserve
export BASE_CERT_DIR=${BASE_DIR}/certs
export DOMAIN_NAME=$(oc get ingresses.config.openshift.io cluster -o jsonpath='{.spec.domain}' | awk -F'.' '{print $(NF-1)"."$NF}')
export COMMON_NAME=$(oc get ingresses.config.openshift.io cluster -o jsonpath='{.spec.domain}'|sed 's/apps.//')

mkdir ${BASE_DIR}
mkdir ${BASE_CERT_DIR}
```

After this new change, should we stop executing wildcard script? 
cc: @Jooho @Xaenalt @heyselbi

**Comment by israel-hdez** [comment]
@bdattoma The solution applied in https://github.com/opendatahub-io/odh-manifests/pull/916 does not prevent creating such certificates.

There is a follow-up issue: #67. With OpenShift Ingress involved, it potentially may remove the need of generating the wildcard certificate. But I'm not sure about the impl. details.

cc @vaibhavjainwiz .

**Comment by israel-hdez** [comment]
> what would I expect to see when reaching that inference service using the OpenShift route?

Before this issue was fixed, the URL stated in the `InferenceService` didn't work -- i.e. the `isvc` is unreachable. You had to look at the `ksvc` and use that URL.

With this issue fixed, the `InferenceService` has the URL that works (i.e. now, it is reachable). You no longer need to worry about the `ksvc`, so it is better. The small hiccup is that the URL is unstable because it can change if depending on the configs of the `isvc` (so 3rd party apps would need to re-read in case of changes). -- FWIW, the KServe features we are using right now won't cause a change in the URL.

Once #67 is implemented, the URL of the `isvc` should become stable regardless of the configuration of the `isvc`.

**Comment by bdattoma** [comment]
> > what would I expect to see when reaching that inference service using the OpenShift route?
> 
> Before this issue was fixed, the URL stated in the `InferenceService` didn't work -- i.e. the `isvc` is unreachable. You had to look at the `ksvc` and use that URL.
> 
> With this issue fixed, the `InferenceService` has the URL that works (i.e. now, it is reachable). You no longer need to worry about the `ksvc`, so it is better. The small hiccup is that the URL is unstable because it can change if depending on the configs of the `isvc` (so 3rd party apps would need to re-read in case of changes). -- FWIW, the KServe features we are using right now won't cause a change in the URL.
> 
> Once #67 is implemented, the URL of the `isvc` should become stable regardless of the configuration of the `isvc`.

ah got it. Looking into my installation, I see isvc and kisvc having the same exact URL, so maybe I've already fetched this change in the build I'm using (i.e., RC2). 

Thus, now in our tests we could fetch url from ISVC. Well, would that hiccup make the product usable in a production environment? What if you use an external application which needs to query the route endpoint? It cannot fetch everytime the new route if the application is served outside the cluster where the model is. Could we give the option to go back to ksvc only?

**Comment by bdattoma** [comment]
ah another question sorry. I see that isvc and ksvc has the same URL now, but I still cannot see the OpenShift route created. Would that happen as part of #67 ?

**Comment by israel-hdez** [comment]
> Well, would that hiccup make the product usable in a production environment?

So far, on my testing, the only reason I've found that would trigger a change in the URL is when [adding/removing a transformer](https://kserve.github.io/website/0.10/modelserving/v1beta1/transformer/feast/). If you only configure a `predictor` it won't change despite any changes you do within the `predictor` field.

I'd say that for our current use-cases, it is OK in production.

> Could we give the option to go back to ksvc only?

The underlying issue won't be solved. If you only configure a `predictor` in the `isvc` you get one corresponding `ksvc`. If you later add a `transformer` in the `isvc`, then two `ksvc` are created: one for the predictor and one for the transformer. The transformer is the one that needs to be the target, so 3rd party apps would still need to switch to that URL. If in a third iteration you remove the `transfomer` then the relevant `ksvc` vanishes and, again, 3rd party apps would need to switch back.

AFAIK, we won't be using transformers for now, so it should be OK for production. In any case, if #67 should be implemented this sprint.

Basically, KServe should be abstracting away the `ksvc` that should be target of traffic behind a fixed/stable URL, and redirecting traffic appropriately.

> I see that isvc and ksvc has the same URL now, but I still cannot see the OpenShift route created. Would that happen as part of https://github.com/opendatahub-io/odh-model-controller/issues/67 ?

Yes, that's the goal.

**Comment by israel-hdez** [comment]
> would this be solved by https://github.com/opendatahub-io/odh-model-controller/issues/67 ?

@bdattoma Yes :slightly_smiling_face:
